// @hirosystems/chainhooks-client/dist/src/schemas/chainhook-event/stacks/block.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksBlockSchema = exports.StacksBlockMetadataSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../common");
const transaction_1 = require("./transaction");
const _1 = require(".");
const utils_1 = require("../../utils");
exports.StacksBlockMetadataSchema = typebox_1.Type.Object({
    canonical: typebox_1.Type.Boolean(),
    burn_block_identifier: common_1.BlockIdentifierSchema,
    burn_block_timestamp: typebox_1.Type.Integer(),
    parent_microblock_identifier: (0, utils_1.Nullable)(common_1.BlockIdentifierSchema),
    tenure_height: typebox_1.Type.Integer(),
    execution_cost: _1.StacksExecutionCostSchema,
    tx_total_size: typebox_1.Type.Integer(),
    tx_count: typebox_1.Type.Integer(),
    signatures: typebox_1.Type.Optional(typebox_1.Type.Object({
        signer_bitvec: (0, utils_1.Nullable)(typebox_1.Type.String()),
        signer_signatures: (0, utils_1.Nullable)(typebox_1.Type.Array(typebox_1.Type.String())),
    })),
    // TODO: Available starting in epoch3, only included in blocks where the pox cycle rewards are first calculated
    // cycle_number: Nullable(Type.Integer()),
    // reward_set: Nullable(
    //   Type.Object({
    //     pox_ustx_threshold: Type.String(),
    //     rewarded_addresses: Type.Array(Type.String()),
    //     signers: Nullable(
    //       Type.Array(
    //         Type.Object({
    //           signing_key: Type.String(),
    //           weight: Type.Integer(),
    //           stacked_amt: Type.String(),
    //         })
    //       )
    //     ),
    //   })
    // ),
});
exports.StacksBlockSchema = typebox_1.Type.Object({
    block_identifier: common_1.BlockIdentifierSchema,
    parent_block_identifier: common_1.BlockIdentifierSchema,
    timestamp: typebox_1.Type.Integer(),
    transactions: typebox_1.Type.Array(transaction_1.StacksTransactionSchema),
    metadata: typebox_1.Type.Optional(exports.StacksBlockMetadataSchema),
});
