// @hirosystems/chainhooks-client/dist/src/schemas/chainhook-event/stacks/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksExecutionCostSchema = void 0;
__exportStar(require("./block"), exports);
__exportStar(require("./transaction"), exports);
__exportStar(require("./operations"), exports);
__exportStar(require("./post_conditions"), exports);
const typebox_1 = require("@sinclair/typebox");
exports.StacksExecutionCostSchema = typebox_1.Type.Object({
    read_count: typebox_1.Type.Integer(),
    read_length: typebox_1.Type.Integer(),
    runtime: typebox_1.Type.Integer(),
    write_count: typebox_1.Type.Integer(),
    write_length: typebox_1.Type.Integer(),
});
