// @hirosystems/chainhooks-client/dist/src/schemas/chainhook/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainhookSchema = exports.ChainhookStatusSchema = exports.ChainhookStatusValueSchema = exports.ChainhookDefinitionSchema = exports.StacksChainhookDefinitionSchema = exports.ChainhookNetworkSchema = void 0;
__exportStar(require("./stacks"), exports);
__exportStar(require("./action"), exports);
const typebox_1 = require("@sinclair/typebox");
const stacks_1 = require("./stacks");
const utils_1 = require("../utils");
const action_1 = require("./action");
const filters_1 = require("./stacks/filters");
exports.ChainhookNetworkSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('mainnet'),
    typebox_1.Type.Literal('testnet'),
]);
exports.StacksChainhookDefinitionSchema = typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    version: typebox_1.Type.Literal('1'),
    chain: typebox_1.Type.Literal('stacks'),
    network: exports.ChainhookNetworkSchema,
    filters: filters_1.StacksFiltersSchema,
    options: typebox_1.Type.Optional((0, utils_1.Nullable)(stacks_1.StacksChainhookOptionsSchema)),
    action: action_1.ChainhookActionSchema,
});
exports.ChainhookDefinitionSchema = typebox_1.Type.Union([
    exports.StacksChainhookDefinitionSchema,
    // TODO: Add Bitcoin chainhook options and conditions. Do not use `Type.Composite` here because it
    // breaks Fastify request validation when registering a new Chainhook.
]);
exports.ChainhookStatusValueSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('new'),
    typebox_1.Type.Literal('streaming'),
    typebox_1.Type.Literal('expired'),
    typebox_1.Type.Literal('interrupted'),
], {
    description: 'Valid chainhook status values: new, streaming, expired, or interrupted',
});
exports.ChainhookStatusSchema = typebox_1.Type.Object({
    status: exports.ChainhookStatusValueSchema,
    enabled: typebox_1.Type.Boolean(),
    created_at: typebox_1.Type.Number(),
    last_evaluated_at: (0, utils_1.Nullable)(typebox_1.Type.Number()),
    last_evaluated_block_height: (0, utils_1.Nullable)(typebox_1.Type.Number()),
    last_occurrence_at: (0, utils_1.Nullable)(typebox_1.Type.Number()),
    last_occurrence_block_height: (0, utils_1.Nullable)(typebox_1.Type.Number()),
    evaluated_block_count: typebox_1.Type.Number(),
    occurrence_count: typebox_1.Type.Number(),
});
exports.ChainhookSchema = typebox_1.Type.Object({
    uuid: typebox_1.Type.String({ format: 'uuid' }),
    definition: exports.ChainhookDefinitionSchema,
    status: exports.ChainhookStatusSchema,
});
