// @hirosystems/chainhooks-client/dist/src/schemas/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainhooksHttpError = void 0;
class ChainhooksHttpError extends Error {
    status;
    url;
    method;
    headers;
    bodyText;
    body;
    name = 'ChainhooksHttpError';
    constructor(status, url, method, headers, bodyText, body) {
        // Include a short preview of the body to keep the message smaller
        const preview = body
            ? JSON.stringify(body).slice(0, 500)
            : bodyText.slice(0, 500);
        super(`ChainhooksClient error: HTTP ${status}${preview ? ` — ${preview}` : ''}`);
        this.status = status;
        this.url = url;
        this.method = method;
        this.headers = headers;
        this.bodyText = bodyText;
        this.body = body;
    }
    toJSON() {
        return {
            name: this.name,
            status: this.status,
            url: this.url,
            method: this.method,
            headers: this.headers,
            bodyText: this.bodyText,
            body: this.body,
            message: this.message,
            stack: this.stack,
        };
    }
}
exports.ChainhooksHttpError = ChainhooksHttpError;
