// @hirosystems/chainhooks-client/dist/src/schemas/chainhook/stacks/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksChainhookOptionsSchema = void 0;
__exportStar(require("./filters"), exports);
const typebox_1 = require("@sinclair/typebox");
exports.StacksChainhookOptionsSchema = typebox_1.Type.Object({
    enable_on_registration: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    expire_after_evaluations: typebox_1.Type.Optional(typebox_1.Type.Integer()),
    expire_after_occurrences: typebox_1.Type.Optional(typebox_1.Type.Integer()),
    decode_clarity_values: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_contract_abi: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_contract_source_code: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_post_conditions: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_raw_transactions: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_block_signatures: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    include_block_metadata: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
}, { additionalProperties: false });
