// @hirosystems/chainhooks-client/schemas/chainhook-event/stacks/operations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksOperationSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../common");
const utils_1 = require("../../utils");
const _1 = require(".");
const StacksCurrencySchema = typebox_1.Type.Object({
    decimals: typebox_1.Type.Integer(),
    symbol: typebox_1.Type.String(),
    metadata: typebox_1.Type.Optional(typebox_1.Type.Object({
        asset_identifier: (0, utils_1.Nullable)(typebox_1.Type.String()),
        value: typebox_1.Type.Optional(typebox_1.Type.String()),
        token_type: typebox_1.Type.Union([typebox_1.Type.Literal('ft'), typebox_1.Type.Literal('nft')]),
    })),
});
const StacksAmountSchema = typebox_1.Type.Object({
    currency: StacksCurrencySchema,
    value: typebox_1.Type.String(),
});
const BaseOperationSchema = typebox_1.Type.Object({
    operation_identifier: common_1.OperationIdentifierSchema,
    status: _1.StacksStatusSchema,
});
const BaseAccountOperationSchema = typebox_1.Type.Composite([
    BaseOperationSchema,
    typebox_1.Type.Object({
        account: common_1.OperationAccountSchema,
    }),
]);
const BaseAmountOperationSchema = typebox_1.Type.Composite([
    BaseAccountOperationSchema,
    typebox_1.Type.Object({
        amount: StacksAmountSchema,
    }),
]);
const CoinbaseOperationSchema = typebox_1.Type.Composite([
    BaseAccountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('coinbase'),
        metadata: typebox_1.Type.Object({
            alt_recipient: (0, utils_1.Nullable)(typebox_1.Type.String()),
            vrf_proof: (0, utils_1.Nullable)(typebox_1.Type.String()),
        }),
    }),
]);
const FeeOperationSchema = typebox_1.Type.Composite([
    BaseAmountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('fee'),
        metadata: typebox_1.Type.Object({
            sponsored: typebox_1.Type.Boolean(),
        }),
    }),
]);
const TokenTransferOperationSchema = typebox_1.Type.Composite([
    BaseAmountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('token_transfer'),
        metadata: typebox_1.Type.Optional(typebox_1.Type.Object({ memo: (0, utils_1.Nullable)(typebox_1.Type.String()) })),
    }),
]);
const TokenMintOperationSchema = typebox_1.Type.Composite([
    BaseAmountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('token_mint'),
    }),
]);
const TokenBurnOperationSchema = typebox_1.Type.Composite([
    BaseAmountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('token_burn'),
    }),
]);
const ContractCallOperationSchema = typebox_1.Type.Composite([
    BaseAccountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('contract_call'),
        metadata: typebox_1.Type.Object({
            args: typebox_1.Type.Union([
                typebox_1.Type.Array(typebox_1.Type.Object({
                    hex: typebox_1.Type.String(),
                    repr: typebox_1.Type.String(),
                    name: typebox_1.Type.String(),
                    type: typebox_1.Type.String(),
                })),
                (0, utils_1.Nullable)(typebox_1.Type.String()),
            ]),
            contract_identifier: typebox_1.Type.String(),
            function_name: typebox_1.Type.String(),
        }),
    }),
]);
const SmartContractOperationSchema = typebox_1.Type.Composite([
    BaseAccountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('contract_deploy'),
        metadata: typebox_1.Type.Object({
            contract_identifier: typebox_1.Type.String(),
            clarity_version: (0, utils_1.Nullable)(typebox_1.Type.Integer()),
            source_code: typebox_1.Type.Optional(typebox_1.Type.String()),
            abi: typebox_1.Type.Optional(typebox_1.Type.String()),
        }),
    }),
]);
const TenureChangeOperationSchema = typebox_1.Type.Composite([
    BaseOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('tenure_change'),
        metadata: typebox_1.Type.Object({
            tenure_consensus_hash: typebox_1.Type.String(),
            prev_tenure_consensus_hash: typebox_1.Type.String(),
            burn_view_consensus_hash: typebox_1.Type.String(),
            previous_tenure_end: typebox_1.Type.String(),
            previous_tenure_blocks: typebox_1.Type.Integer(),
            cause: typebox_1.Type.Union([
                typebox_1.Type.Literal('block_found'),
                typebox_1.Type.Literal('extended'),
            ]),
            pubkey_hash: typebox_1.Type.String(),
        }),
    }),
]);
const PoisonMicroblockOperationSchema = typebox_1.Type.Composite([
    BaseAccountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('poison_microblock'),
        metadata: typebox_1.Type.Object({
            microblock_header_1: typebox_1.Type.String(),
            microblock_header_2: typebox_1.Type.String(),
        }),
    }),
]);
const StxLockOperationSchema = typebox_1.Type.Composite([
    BaseAmountOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('stx_lock'),
        metadata: typebox_1.Type.Object({
            unlock_height: typebox_1.Type.Integer(),
            locked_address: typebox_1.Type.String(),
            contract_name: typebox_1.Type.String(),
        }),
    }),
]);
const ContractLogOperationSchema = typebox_1.Type.Composite([
    BaseOperationSchema,
    typebox_1.Type.Object({
        type: typebox_1.Type.Literal('contract_log'),
        metadata: typebox_1.Type.Object({
            contract_identifier: typebox_1.Type.String(),
            topic: typebox_1.Type.String(),
            value: typebox_1.Type.Union([
                typebox_1.Type.String(),
                typebox_1.Type.Object({
                    hex: typebox_1.Type.String(),
                    repr: typebox_1.Type.String(),
                }),
            ]),
        }),
    }),
]);
exports.StacksOperationSchema = typebox_1.Type.Union([
    CoinbaseOperationSchema,
    FeeOperationSchema,
    TokenTransferOperationSchema,
    ContractCallOperationSchema,
    SmartContractOperationSchema,
    TenureChangeOperationSchema,
    PoisonMicroblockOperationSchema,
    TokenMintOperationSchema,
    TokenBurnOperationSchema,
    StxLockOperationSchema,
    ContractLogOperationSchema,
]);
