// @hirosystems/chainhooks-client/schemas/chainhook-event/stacks/post_conditions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksPostConditionSchema = exports.StacksPostConditionModeSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
exports.StacksPostConditionModeSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('allow'),
    typebox_1.Type.Literal('deny'),
]);
const StacksFungiblePostConditionCodeSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('sent_equal_to'),
    typebox_1.Type.Literal('sent_greater_than'),
    typebox_1.Type.Literal('sent_greater_than_or_equal_to'),
    typebox_1.Type.Literal('sent_less_than'),
    typebox_1.Type.Literal('sent_less_than_or_equal_to'),
]);
const StacksNonFungiblePostConditionCodeSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('sent'),
    typebox_1.Type.Literal('not_sent'),
]);
const StacksPostConditionPrincipalStandardSchema = typebox_1.Type.Object({
    type_id: typebox_1.Type.Literal('principal_standard'),
    address: typebox_1.Type.String(),
});
const StacksPostConditionPrincipalContractSchema = typebox_1.Type.Object({
    type_id: typebox_1.Type.Literal('principal_contract'),
    contract_name: typebox_1.Type.String(),
    address: typebox_1.Type.String(),
});
const StacksPostConditionPrincipalOriginSchema = typebox_1.Type.Object({
    type_id: typebox_1.Type.Literal('principal_origin'),
});
const StacksPostConditionPrincipalSchema = typebox_1.Type.Union([
    StacksPostConditionPrincipalStandardSchema,
    StacksPostConditionPrincipalContractSchema,
    StacksPostConditionPrincipalOriginSchema,
]);
const StacksPostConditionAssetSchema = typebox_1.Type.Object({
    contract_name: typebox_1.Type.String(),
    asset_name: typebox_1.Type.String(),
    contract_address: typebox_1.Type.String(),
});
const StacksStxPostConditionSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('stx'),
    condition_code: StacksFungiblePostConditionCodeSchema,
    amount: typebox_1.Type.String(),
    principal: StacksPostConditionPrincipalSchema,
});
const StacksFungiblePostConditionSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('fungible'),
    condition_code: StacksFungiblePostConditionCodeSchema,
    amount: typebox_1.Type.String(),
    principal: StacksPostConditionPrincipalSchema,
    asset: StacksPostConditionAssetSchema,
});
const StacksNonFungiblePostConditionSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('non_fungible'),
    condition_code: StacksNonFungiblePostConditionCodeSchema,
    principal: StacksPostConditionPrincipalSchema,
    asset: StacksPostConditionAssetSchema,
    asset_value: typebox_1.Type.Object({
        hex: typebox_1.Type.String(),
        repr: typebox_1.Type.String(),
    }),
});
exports.StacksPostConditionSchema = typebox_1.Type.Union([
    StacksStxPostConditionSchema,
    StacksFungiblePostConditionSchema,
    StacksNonFungiblePostConditionSchema,
]);
