// @hirosystems/chainhooks-client/schemas/chainhook-event/stacks/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksTransactionSchema = exports.StacksTransactionMetadataSchema = exports.StacksTransactionPositionSchema = exports.StacksTransactionTypeSchema = exports.StacksStatusSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../common");
const _1 = require(".");
const utils_1 = require("../../utils");
const operations_1 = require("./operations");
const post_conditions_1 = require("./post_conditions");
exports.StacksStatusSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('pending'),
    typebox_1.Type.Literal('success'),
    typebox_1.Type.Literal('abort_by_response'),
    typebox_1.Type.Literal('abort_by_post_condition'),
    typebox_1.Type.Literal('dropped_replace_by_fee'),
    typebox_1.Type.Literal('dropped_replace_across_fork'),
    typebox_1.Type.Literal('dropped_too_expensive'),
    typebox_1.Type.Literal('dropped_problematic'),
    typebox_1.Type.Literal('dropped_stale_garbage_collect'),
]);
exports.StacksTransactionTypeSchema = typebox_1.Type.Union([
    typebox_1.Type.Literal('coinbase'),
    typebox_1.Type.Literal('token_transfer'),
    typebox_1.Type.Literal('poison_microblock'),
    typebox_1.Type.Literal('contract_deploy'),
    typebox_1.Type.Literal('tenure_change'),
    typebox_1.Type.Literal('contract_call'),
]);
exports.StacksTransactionPositionSchema = typebox_1.Type.Object({
    index: typebox_1.Type.Integer(),
    microblock_identifier: (0, utils_1.Nullable)(common_1.BlockIdentifierSchema),
});
exports.StacksTransactionMetadataSchema = typebox_1.Type.Object({
    canonical: typebox_1.Type.Boolean(),
    microblock_canonical: typebox_1.Type.Boolean(),
    execution_cost: _1.StacksExecutionCostSchema,
    fee_rate: typebox_1.Type.String(),
    nonce: typebox_1.Type.Integer(),
    type: exports.StacksTransactionTypeSchema,
    post_conditions: typebox_1.Type.Optional(typebox_1.Type.Object({
        mode: post_conditions_1.StacksPostConditionModeSchema,
        post_conditions: typebox_1.Type.Array(post_conditions_1.StacksPostConditionSchema),
    })),
    position: exports.StacksTransactionPositionSchema,
    raw_tx: typebox_1.Type.Optional(typebox_1.Type.String()),
    result: typebox_1.Type.Union([
        typebox_1.Type.String(),
        typebox_1.Type.Object({
            hex: typebox_1.Type.String(),
            repr: typebox_1.Type.String(),
        }),
    ]),
    sender_address: typebox_1.Type.String(),
    sponsor_address: (0, utils_1.Nullable)(typebox_1.Type.String()),
    sponsored: typebox_1.Type.Boolean(),
    status: exports.StacksStatusSchema,
    vm_error: (0, utils_1.Nullable)(typebox_1.Type.String()),
});
exports.StacksTransactionSchema = typebox_1.Type.Object({
    transaction_identifier: common_1.TransactionIdentifierSchema,
    operations: typebox_1.Type.Array(operations_1.StacksOperationSchema),
    metadata: exports.StacksTransactionMetadataSchema,
});
