// @hirosystems/chainhooks-client/schemas/chainhook/stacks/filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksFiltersSchema = exports.StacksEventSchema = exports.StacksEventTenureChangeSchema = exports.StacksEventCoinbaseSchema = exports.StacksEventContractLogSchema = exports.StacksEventContractCallSchema = exports.StacksEventSmartContractSchema = exports.StacksEventStxTransferSchema = exports.StacksEventStxBurnSchema = exports.StacksEventStxMintSchema = exports.StacksEventStxEventSchema = exports.StacksEventNftTransferSchema = exports.StacksEventNftBurnSchema = exports.StacksEventNftMintSchema = exports.StacksEventNftEventSchema = exports.StacksEventFtTransferSchema = exports.StacksEventFtBurnSchema = exports.StacksEventFtMintSchema = exports.StacksEventFtEventSchema = exports.StacksEventBalanceChangeSchema = exports.StacksAssetIdentifierSchema = exports.StacksPrincipalSchema = exports.StacksAddressSchema = exports.StacksContractIdentifierSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
exports.StacksContractIdentifierSchema = typebox_1.Type.String({
    pattern: '^[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41}\\.[a-zA-Z]([a-zA-Z0-9]|[-_]){0,39}$',
});
exports.StacksAddressSchema = typebox_1.Type.String({
    pattern: '^[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41}$',
});
exports.StacksPrincipalSchema = typebox_1.Type.Union([
    exports.StacksContractIdentifierSchema,
    exports.StacksAddressSchema,
]);
exports.StacksAssetIdentifierSchema = typebox_1.Type.String({
    pattern: '^[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41}\\.[a-zA-Z]([a-zA-Z0-9]|[-_]){0,39}::[a-zA-Z]([a-zA-Z0-9]|[-_]){0,39}$',
});
exports.StacksEventBalanceChangeSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('balance_change'),
    principal: exports.StacksPrincipalSchema,
});
exports.StacksEventFtEventSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('ft_event'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
});
exports.StacksEventFtMintSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('ft_mint'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventFtBurnSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('ft_burn'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventFtTransferSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('ft_transfer'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventNftEventSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('nft_event'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    value: typebox_1.Type.Optional(typebox_1.Type.Integer()),
});
exports.StacksEventNftMintSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('nft_mint'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    value: typebox_1.Type.Optional(typebox_1.Type.Integer()),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventNftBurnSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('nft_burn'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    value: typebox_1.Type.Optional(typebox_1.Type.Integer()),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventNftTransferSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('nft_transfer'),
    asset_identifier: typebox_1.Type.Optional(exports.StacksAssetIdentifierSchema),
    value: typebox_1.Type.Optional(typebox_1.Type.Integer()),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventStxEventSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('stx_event'),
});
exports.StacksEventStxMintSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('stx_mint'),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventStxBurnSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('stx_burn'),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventStxTransferSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('stx_transfer'),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
    receiver: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventSmartContractSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('contract_deploy'),
    contract_identifier: typebox_1.Type.Optional(exports.StacksContractIdentifierSchema),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventContractCallSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('contract_call'),
    contract_identifier: typebox_1.Type.Optional(exports.StacksContractIdentifierSchema),
    function_name: typebox_1.Type.Optional(typebox_1.Type.String()),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventContractLogSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('contract_log'),
    contract_identifier: typebox_1.Type.Optional(exports.StacksContractIdentifierSchema),
    sender: typebox_1.Type.Optional(exports.StacksPrincipalSchema),
});
exports.StacksEventCoinbaseSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('coinbase'),
});
exports.StacksEventTenureChangeSchema = typebox_1.Type.Object({
    type: typebox_1.Type.Literal('tenure_change'),
    cause: typebox_1.Type.Union([typebox_1.Type.Literal('block_found'), typebox_1.Type.Literal('extended')]),
});
exports.StacksEventSchema = typebox_1.Type.Union([
    exports.StacksEventBalanceChangeSchema,
    exports.StacksEventFtEventSchema,
    exports.StacksEventFtMintSchema,
    exports.StacksEventFtBurnSchema,
    exports.StacksEventFtTransferSchema,
    exports.StacksEventNftEventSchema,
    exports.StacksEventNftMintSchema,
    exports.StacksEventNftBurnSchema,
    exports.StacksEventNftTransferSchema,
    exports.StacksEventStxEventSchema,
    exports.StacksEventStxMintSchema,
    exports.StacksEventStxBurnSchema,
    exports.StacksEventStxTransferSchema,
    exports.StacksEventSmartContractSchema,
    exports.StacksEventContractCallSchema,
    exports.StacksEventContractLogSchema,
    exports.StacksEventCoinbaseSchema,
    exports.StacksEventTenureChangeSchema,
]);
exports.StacksFiltersSchema = typebox_1.Type.Object({
    events: typebox_1.Type.Array(exports.StacksEventSchema),
});
